lib/state/analytics/middleware.ts

```diff
@@ -0,0 +1,82 @@
+import analytics from '../../analytics';
+import actions from '../actions';
+
+import * as A from '../action-types';
+import * as S from '../';
+import * as T from '../../types';
+
+export const middleware: S.Middleware = store => {
+  return next => (action: A.ActionType) => {
+    const result = next(action);
+    const nextState = store.getState();
+
+    // @todo events throughout the app that don't have a corresponding action
+    // analytics.tracks.recordEvent('user_account_created');
+    // analytics.tracks.recordEvent('editor_checklist_inserted');
+    // analytics.tracks.recordEvent('importer_import_completed', {
+    // analytics.tracks.recordEvent('editor_note_collaborator_added');
+    // analytics.tracks.recordEvent('editor_note_collaborator_removed');
+    // analytics.tracks.recordEvent('editor_note_edited');
+    // analytics.tracks.recordEvent('list_tag_deleted');
+
+    switch (action.type) {
+      case 'AUTH_SET':
+        if (action.status === 'authorized') {
+          // @todo called twice for some reason
+          analytics.tracks.recordEvent('user_signed_in');
+        }
+        // @todo this might need to be App.setAccountName
+        // if (action.status === 'not-authorized') {
+        //   analytics.tracks.recordEvent('user_signed_out');
+        // }
+        break;
+      case 'App.loadPreferences':
+        // @todo not getting called
+        analytics.tracks.recordEvent('application_opened');
+        break;
+      // case 'DELETE_NOTE_FOREVER': @todo not sure if?
+      case 'TRASH_NOTE':
+        analytics.tracks.recordEvent('editor_note_deleted');
+        break;
+      case 'CREATE_NOTE':
+        analytics.tracks.recordEvent('list_note_created');
+        break;
+      case 'OPEN_TAG':
+        analytics.tracks.recordEvent('list_tag_viewed');
+        break;
+      case 'RESTORE_NOTE':
+        analytics.tracks.recordEvent('editor_note_restored');
+        break;
+      case 'SEARCH':
+        if (action.searchQuery) {
+          // @todo this gets called for each character typed,
+          // possibly also happening in develop?
+          analytics.tracks.recordEvent('list_notes_searched');
+        }
```

Comment:
we can trap `FILTER_NOTES` instead if we prefer. we could either get the search query from state or we could add the searched-for query into the `FILTER_NOTES` action itself

Reply:
FILTER_NOTES seems even worse, as it gets called in lots of places not triggered by searching 🤔

Reply:
ah okay. we're specifically trying to see every time someone enters a character into the search field…

```
s
so
som
some
somet
someth
somethi
somethin
something
```

this could be a decent place for a debounce possibly unless we want the numbers to be grossly inflated, which of course I'm guessing they already are…

Reply:
yeah, that was my "possibly also happening in develop?" comment. I am pretty sure this is, intentionally or not, what this event is already tracking